defender/staged/downloaded/Compound_Gasless_Voting.js

```diff
@@ -0,0 +1,122 @@
+  const provider = new DefenderRelayProvider(autotaskEvent);
+
+  // initiate a new relay signer
+  console.debug('Creating DefenderRelaySigner');
+  const signer = new DefenderRelaySigner(autotaskEvent, provider, { speed: 'fast' });
+
+  // Test relay and check for balance
+  let relayAddress;
+  try {
+    relayAddress = await signer.getAddress();
+  } catch (error) {
+    console.error('Relay is not working, check if it is connected : ', error);
+    throw error;
+  }
+
+  const relayBalance = await provider.getBalance(relayAddress);
+  if (relayBalance <= 0) {
+    throw new Error('Insufficient funds for Relay, please deposit funds.')
+  }
+
+  // create an ethers.js Contract Object to interact with the on-chain governance smart contract
+  console.debug('Creating governanceContract');
+  const governanceContract = new ethers.Contract(
+    compoundGovernanceAddress,
+    compoundGovernanceAbi,
+    signer,
+  );
+
+  // create an ethers.js Contract Object to interact with the on-chain COMP smart contract
+  console.debug('Creating compContract');
+  const compTokenContract = new ethers.Contract(
+    compTokenAddress,
+    compTokenAbi,
+    signer,
+  );
+
+  // cast vote or delegate vote on-chain
+  const promises = body.map(async (votes) => {
+    if (votes.delegatee !== undefined) {
+      const {
+        address, delegatee, nonce, expiry, v, r, s,
+      } = votes;
+
+      // check if address has votes to delegate
+      const delegateCompBalance = await compTokenContract.balanceOf(address);
+      if (delegateCompBalance > 0) {
+        // delegate vote on-chain
+        console.debug(`Delgating votes from ${address} to ${delegatee}`);
+        try {
+          await compTokenContract.delegateBySig(delegatee, nonce, expiry, v, r, s);
+        } catch (error) {
+          console.error(error);
+        }
+      }
+    } else if (votes.support !== undefined) {
+      const {
+        address, proposalId, support, v, r, s,
+      } = votes;
+
+      // check if address has enough delegated votes to submit a vote
+      // check if proposal is active - only active proposals can be voted on
+      const addressDelegatedVotes = await compTokenContract.getCurrentVotes(address);
+      const state = await governanceContract.state(proposalId);
+      if (addressDelegatedVotes > 0 && state === 1) {
+        if (support === 0) {
+          console.debug(`Address ${address} is casting a vote in favor of proposal ID: ${proposalId}`);
+        } else if (support === 1) {
+          console.debug(`Address ${address} is casting a vote against proposal ID: ${proposalId}`);
+        } else if (support === 2) {
+          console.debug(`Address ${address} is casting a vote to abstain for proposal ID: ${proposalId}`);
+        }
+        // cast vote on-chain
+        try {
+          await governanceContract.castVoteBySig(proposalId, support, v, r, s);
+        } catch (error) {
+          console.error(error);
```

Comment:
If the `castVoteBySig` was unsuccessful, this should throw an error.

Reply:
Since you are processing a batch of requests, it might be better to store the errors and throw them after everything else is processed.

Reply:
Agreed. You can remove the `try...catch`, replace the `await Promise.all(promises)` with `const results = await Promise.allSettled(promises)`, then check the `status` of each settled Promise.

If any of the Promises is rejected, you could throw an Error in the Autotask with the combined reasons (each rejected Promise will have a reason string), or use `console.error()` with it and return false to the caller, depending upon who you want to notify.

My guess is that you don't want the Autotask owner to receive Defender emails when a caller fails to use the Autotask correctly. You probably just want a record of the failed call in the Defender logs and then an appropriate response returned to the caller.

Ref: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled
Example: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled#examples

Reply:
removed `try catch` and used `Promise.allSettled()` and returned `false` if there is a failed transaction